impeller/display_list/dl_aiks_canvas.cc

```diff
@@ -0,0 +1,1191 @@
+      if (!color_filter) {
+        return nullptr;
+      }
+      return [color_filter](FilterInput::Ref input,
+                            const Matrix& effect_transform, bool is_subpass) {
+        // When color filters are used as image filters, set the color filter's
+        // "absorb opacity" flag to false. For image filters, the snapshot
+        // opacity needs to be deferred until the result of the filter chain is
+        // being blended with the layer.
+        return color_filter->WrapWithGPUColorFilter(std::move(input), false);
+      };
+      break;
+    }
+    case flutter::DlImageFilterType::kLocalMatrix: {
+      auto local_matrix_filter = filter->asLocalMatrix();
+      FML_DCHECK(local_matrix_filter);
+      auto internal_filter = local_matrix_filter->image_filter();
+      FML_DCHECK(internal_filter);
+
+      auto image_filter_proc = ToImageFilterProc(internal_filter.get());
+      if (!image_filter_proc) {
+        return nullptr;
+      }
+
+      auto matrix = ToMatrix(local_matrix_filter->matrix());
+
+      return [matrix, filter_proc = image_filter_proc](
+                 FilterInput::Ref input, const Matrix& effect_transform,
+                 bool is_subpass) {
+        std::shared_ptr<FilterContents> filter =
+            filter_proc(std::move(input), effect_transform, is_subpass);
+        return FilterContents::MakeLocalMatrixFilter(FilterInput::Make(filter),
+                                                     matrix);
+      };
+      break;
+    }
+  }
+}
+
+static Paint ToPaint(const flutter::DlPaint& dl_paint) {
+  Paint paint;
+  paint.style = ToStyle(dl_paint.getDrawStyle());
+  paint.color = ToColor(dl_paint.getColor());
+  paint.stroke_width = dl_paint.getStrokeWidth();
+  paint.stroke_miter = dl_paint.getStrokeMiter();
+  paint.stroke_cap = ToStrokeCap(dl_paint.getStrokeCap());
+  paint.stroke_join = ToStrokeJoin(dl_paint.getStrokeJoin());
+  paint.color_source = ToColorSource(dl_paint.getColorSourcePtr(), &paint);
+  paint.color_filter = ToColorFilter(dl_paint.getColorFilterPtr());
+  paint.invert_colors = dl_paint.isInvertColors();
+  paint.blend_mode = ToBlendMode(dl_paint.getBlendMode());
+  if (dl_paint.getPathEffect()) {
+    UNIMPLEMENTED;
+  }
+  paint.mask_blur_descriptor =
+      ToMaskBlurDescriptor(dl_paint.getMaskFilterPtr());
+  paint.image_filter = ToImageFilterProc(dl_paint.getImageFilterPtr());
+  return paint;
+}
+
+static Paint ToPaint(const flutter::DlPaint* dl_paint) {
+  if (!dl_paint) {
+    return Paint();
+  }
+  return ToPaint(*dl_paint);
+}
+
+// |flutter::DlCanvas|
+void DlAiksCanvas::Save() {
+  canvas_.Save();
+  if (accumulator_) {
+    accumulator_->save();
+  }
+}
+
+// |flutter::DlCanvas|
+void DlAiksCanvas::SaveLayer(const SkRect* bounds,
+                             const flutter::DlPaint* paint,
+                             const flutter::DlImageFilter* backdrop) {
+  if (!paint) {
+    return;
```

Comment:
This is a bug. Paint is optional, you can't just return. Perhaps we don't end up here with paint being null, but it doesn't invalidate the operation, it just means that the collected pixels are applied with default paint values.